### ISSUE TITLE
Help with yolov8 segmentation labels

### ISSUE THREAD

**Issue by AyushExel** [akashAD98/YOLOV8_SAM#2]
@laughing-q I met with akash today and we have a nice idea about creating an autolabeler for yolov8 based on this work.

@akashAD98 can you point to the part of the code that is causing problems? I'll take a look and try to reproduce. Also let me know if you need any help with writing your first PR. Happy to assist you/

**Comment by Laughing-q** [comment]
@akashAD98 here's how I save and plot masks, looks good to me.
save:
```python
from ultralytics import YOLO
import numpy as np
import cv2
import torch
from segment_anything import sam_model_registry, SamPredictor

def yolov8_detection(model, image_path):

    image = cv2.imread(image_path)
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    results = model(image, stream=True)  # generator of Results objects

    for result in results:
        boxes = result.boxes  # Boxes object for bbox outputs

    bbox = boxes.xyxy.tolist()
    bbox = [[int(i) for i in box] for box in bbox]
    return bbox, image


model = YOLO("yolov8n.pt")
yolov8_boxex, image = yolov8_detection(model, "/home/laughing/codes/ultralytics/ultralytics/assets/bus.jpg")
input_boxes = torch.tensor(yolov8_boxex, device=model.device)

sam_checkpoint = "weights/sam_vit_l_0b3195.pth"
model_type = "vit_l"
device = "cuda:0"

sam = sam_model_registry[model_type](checkpoint=sam_checkpoint)
sam.to(device=device)

predictor = SamPredictor(sam)
predictor.set_image(image)

transformed_boxes = predictor.transform.apply_boxes_torch(input_boxes, image.shape[:2])
masks, _, _ = predictor.predict_torch(
    point_coords=None,
    point_labels=None,
    boxes=transformed_boxes,
    multimask_output=False,
)

for i, mask in enumerate(masks):

    binary_mask = masks[i].squeeze().cpu().numpy().astype(np.uint8)

    # Find the contours of the mask
    contours, hierarchy = cv2.findContours(binary_mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

    largest_contour = max(contours, key=cv2.contourArea)

    # Get the new bounding box
    bbox = [int(x) for x in cv2.boundingRect(largest_contour)]

    # Get the segmentation mask for object
    segmentation = largest_contour.flatten().tolist()

    # Write bounding boxes to file in YOLO format
    with open("BBOX_yolo.txt", "w") as f:
        for contour in contours:
            # Get the bounding box coordinates of the contour
            x, y, w, h = cv2.boundingRect(contour)
            # Convert the coordinates to YOLO format and write to file
            f.write(
                "0 {:.6f} {:.6f} {:.6f} {:.6f}\n".format(
                    (x + w / 2) / image.shape[1],
                    (y + h / 2) / image.shape[0],
                    w / image.shape[1],
                    h / image.shape[0],
                )
            )
            f.write("\n")
    mask = segmentation

    # load the image
    # width, height = image_path.size
    height, width = image.shape[:2]

    # convert mask to numpy array of shape (N,2)
    mask = np.array(mask).reshape(-1, 2)

    # normalize the pixel coordinates
    mask_norm = mask / np.array([width, height])

    # compute the bounding box
    xmin, ymin = mask_norm.min(axis=0)
    xmax, ymax = mask_norm.max(axis=0)
    bbox_norm = np.array([xmin, ymin, xmax, ymax])

    # concatenate bbox and mask to obtain YOLO format
    # yolo = np.concatenate([bbox_norm, mask_norm.reshape(-1)])
    yolo = mask_norm.reshape(-1)

    # compute the bounding box
    # write the yolo values to a text file
    with open("yolomask_format.txt", "a") as f:
        for val in yolo:
            f.write("{:.6f} ".format(val))
        f.write("\n")

    print("Bounding box:", bbox)
    print("yolo", yolo)
```
plot:
```python
import cv2
import numpy as np

image = cv2.imread("/home/laughing/codes/ultralytics/ultralytics/assets/bus.jpg")
h, w = image.shape[:2]
with open("./yolomask_format.txt") as f:
    segment = [np.array(x.split(), dtype=np.float32).reshape(-1, 2) for x in f.read().strip().splitlines() if len(x)]
for s in segment:
    s[:, 0] *= w
    s[:, 1] *= h
cv2.drawContours(image, [s.astype(np.int32) for s in segment], -1, (0, 0, 255))
cv2.imwrite("output.jpg", image)
```
here's my result:
![1](https://user-images.githubusercontent.com/61612323/233521741-6016fd5d-1182-40ac-aeff-5f1c1954dc26.jpg)

**Comment by Laughing-q** [comment]
@akashAD98 back to this comment [https://github.com/ultralytics/ultralytics/issues/2075#issuecomment-1516047158](https://github.com/ultralytics/ultralytics/issues/2075#issuecomment-1516047158), if I also save bbox info:
```python
    yolo = np.concatenate([bbox_norm, mask_norm.reshape(-1)])  # save both  box and mask
    # yolo = mask_norm.reshape(-1)     # mask only
```
 then the plotting looks like:
![1](https://user-images.githubusercontent.com/61612323/233521978-afa7cac0-1d1e-4f5a-ae8d-ff1b43df3c2e.jpg)

**Comment by Laughing-q** [comment]
@akashAD98 I tested the generated label by training a segmentation model, the `train_batch*.jpg` also looks correct to me.
![train_batch0](https://user-images.githubusercontent.com/61612323/233523276-08021e5c-ea1f-4cfe-b4cd-4e80f80a46dc.jpg)

**Comment by akashAD98** [comment]
@AyushExel @Laughing-q thanks a lot for your help. I really love to do contribution in YOLO & ultralytics

<img width="647" alt="image" src="https://user-images.githubusercontent.com/62583018/233530550-74206901-d61e-43ed-bae5-ef858d01721a.png">

so the first & last point was not touching,so that's not a problem, right?

**Comment by Laughing-q** [comment]
@akashAD98 yeah I don't think that matters, but to ensure that you can launch a train with this image and its label to see if `train_batch*.jpg` looks good. :)

**Comment by akashAD98** [comment]
@Laughing-q got it thanks. 
@AyushExel i want to contribute this to ultralytics YOLO, so people can use their existing YOLO models & create mask annotations easily,& they can train models with YOLO

**Comment by AyushExel** [comment]
Hey, Akash just woke up. Could you start a PR with your current work under the file ultralytics/datasets/annotate.py? That way you'll get the credit for making the PR and we can help you with other steps

**Comment by AyushExel** [comment]
Oh btw, ideally we would want to port SAM over to the repo and not just add it as a dependency, which might be a bigger project in itself.

**Comment by akashAD98** [comment]
@AyushExel got it

**Comment by swaroop116** [comment]
Hello, I was wondering if we can add labels from yolov8, on top of bounding boxes in the end result after getting results from SAM. Writing the detected object on top of each bounding box

**Comment by akashAD98** [comment]
Yes i added that support, you can directly get class id & bbox results in
results & visulise it with any tool

**Comment by AyushExel** [comment]
@akashAD98 I've added support for auto annotation in Ultralytics

**Comment by akashAD98** [comment]
@AyushExel thats great .its really helpful for community. 
also is there any way to do contribution there in ultrlytics? regarding this code

**Comment by AyushExel** [comment]
@akashAD98 Yeah sure. Do you think you can improve upon this https://github.com/ultralytics/ultralytics/blob/main/ultralytics/yolo/data/annotator.py ?
As I discussed with you on the call, just make a PR with whatever improvement you want and we'll merge it once approved and give proper credit :)
